examples/paper/03-dependent-binding.makam

```diff
@@ -1,90 +1,216 @@
 %use "02-binding-forms".
 
-(* Sneak peek: work-in-progress and not much commentary here, just roughly showing what's possible. *)
+(* The type system of λProlog can be viewed as a particular subset of System Fω: namely, it is the
+   simply-typed lambda calculus extended with prenex polymorphism and simple type constructors of
+   the form `type -> type -> ... -> type`. (As an aside, `prop` can be viewed as a separate sort,
+   but we take the view that it is just a distinguished extensible type.)
+
+   As is well-known from Haskell even before the addition of kind definitions, type promotion and
+   type-in-type, this subset of System Fω is enough to model some form of dependency. For example,
+   we can introduce two types for modelling natural numbers, and define vectors as a GADT using those:
+*)
+natZ : type.
+natS : type -> type.
+
+vector : type -> type -> type.
+vnil : vector natZ A.
+vcons : A -> vector N A -> vector (natS N) A.
+
+(* In fact, λProlog naturally supports pattern-matching over such constructors as well, through
+   *ad-hoc polymorphism*, where polymorphic type variables are allowed to be instantiated at *runtime*
+   rather than at type-checking time. The mechanism through which ad-hoc polymorphism works in
+   λProlog is simple: before performing unification at the term-level, we perform unification at
+   the type level first, therefore further determining any uninstantiated type variables. Therefore,
+   when we check to see whether the current goal matches the premise of a rule, type unification
+   can force us to distinguish between different types. Based on these, the standard example
+   of `map` for vectors is as follows:
+*)
+
+vmap : [A B] (A -> B -> prop) -> vector N A -> vector N B -> prop.
+vmap P vnil vnil.
+vmap P (vcons X XS) (vcons Y YS) :- P X Y, vmap P XS YS.
+
+(* The notation [N] in the type of `vmap` means that the type argument `N` is ad-hoc/not-parametric.
```

Comment:
(This explanation is not accurate, until I flip the meaning of square bracket notation in types.)